File: python/tensor4all/src/tensor4all/quanticstransform.py
"""QuanticsTransform - Quantics transformation operators.

Provides shift, flip, Fourier, phase rotation, and cumulative sum
operators for quantics tensor train representations.
"""

from __future__ import annotations
from enum import IntEnum

from ._ffi import ffi
from ._capi import get_lib, check_status
from .treetn import TreeTensorNetwork


class BoundaryCondition(IntEnum):
    PERIODIC = 0
    OPEN = 1


class LinearOperator:
    """A linear operator (MPO) for quantics transformations."""

    def __init__(self, ptr):
        if ptr == ffi.NULL:
            raise ValueError("Cannot create LinearOperator from null pointer")
        self._ptr = ptr

    def __del__(self):
        if hasattr(self, "_ptr") and self._ptr != ffi.NULL:
            get_lib().t4a_linop_release(self._ptr)
            self._ptr = ffi.NULL

    def apply(self, state: TreeTensorNetwork, *,
              method: str = "naive", rtol: float = 0.0,
              maxdim: int = 0) -> TreeTensorNetwork:
        """Apply operator to a TreeTensorNetwork (MPS).

        Parameters
        ----------
        state : TreeTensorNetwork
            Input MPS/TreeTN.
        method : str
            "naive", "zipup", or "fit".
        rtol : float
            Relative tolerance (0.0 = default).
        maxdim : int
            Maximum bond dimension (0 = unlimited).
        """
        lib = get_lib()
        method_int = {"naive": 0, "zipup": 1, "fit": 2}.get(method)
        if method_int is None:
            raise ValueError(f"Unknown method: {method}")

        out = ffi.new("t4a_treetn**")
        status = lib.t4a_linop_apply(
            self._ptr, state._handle, method_int, rtol, maxdim, out
        )
        check_status(status, "apply")

        # Get number of vertices for node names
        n_out = ffi.new("size_t*")
        status = lib.t4a_treetn_num_vertices(out[0], n_out)
        check_status(status, "num_vertices")
        n = n_out[0]

        return TreeTensorNetwork._from_handle(out[0], list(range(n)))

    def __repr__(self):
        return "LinearOperator()"


def shift_operator(r: int, offset: int, *,
                   bc: BoundaryCondition = BoundaryCondition.PERIODIC) -> LinearOperator:
    """Create shift operator: f(x) = g(x + offset) mod 2^r.

    Parameters
    ----------
    r : int
        Number of quantics bits.
    offset : int
        Shift offset (can be negative).
    bc : BoundaryCondition
        Boundary condition (PERIODIC or OPEN).
    """
    lib = get_lib()
    out = ffi.new("t4a_linop**")
    status = lib.t4a_qtransform_shift(r, offset, int(bc), out)
    check_status(status, "shift_operator")
    return LinearOperator(out[0])


def flip_operator(r: int, *,
                  bc: BoundaryCondition = BoundaryCondition.PERIODIC) -> LinearOperator:
    """Create flip operator: f(x) = g(2^r - x).

    Parameters
    ----------
    r : int
        Number of quantics bits.
    bc : BoundaryCondition
        Boundary condition (PERIODIC or OPEN).
    """
    lib = get_lib()
    out = ffi.new("t4a_linop**")
    status = lib.t4a_qtransform_flip(r, int(bc), out)
    check_status(status, "flip_operator")
    return LinearOperator(out[0])


def phase_rotation_operator(r: int, theta: float) -> LinearOperator:
    """Create phase rotation operator: f(x) = exp(i*theta*x) * g(x).

    Parameters
    ----------
    r : int
        Number of quantics bits.
    theta : float
        Phase angle.
    """
    lib = get_lib()
    out = ffi.new("t4a_linop**")
    status = lib.t4a_qtransform_phase_rotation(r, theta, out)
    check_status(status, "phase_rotation_operator")
    return LinearOperator(out[0])


def cumsum_operator(r: int) -> LinearOperator:
    """Create cumulative sum operator: y_i = sum_{j<i} x_j.

    Parameters
    ----------
    r : int
        Number of quantics bits.
    """
    lib = get_lib()
    out = ffi.new("t4a_linop**")
    status = lib.t4a_qtransform_cumsum(r, out)
    check_status(status, "cumsum_operator")
    return LinearOperator(out[0])


def fourier_operator(r: int, *, forward: bool = True,
                     maxbonddim: int = 0,
                     tolerance: float = 0.0) -> LinearOperator:
    """Create Quantics Fourier Transform operator.

    Parameters
    ----------
    r : int
        Number of quantics bits.
    forward : bool
        True for forward FT, False for inverse.
    maxbonddim : int
        Maximum bond dimension (0 = default=12).
    tolerance : float
        Construction tolerance (0.0 = default=1e-14).
    """
    lib = get_lib()
    out = ffi.new("t4a_linop**")
    status = lib.t4a_qtransform_fourier(r, 1 if forward else 0, maxbonddim, tolerance, out)
    check_status(status, "fourier_operator")
    return LinearOperator(out[0])

File: julia/Tensor4all.jl/src/QuanticsTransform.jl
"""
    QuanticsTransform

Quantics transformation operators for tree tensor networks (MPS).

Provides operators for shift, flip, phase rotation, cumulative sum, and
Fourier transform in quantics representation, wrapping the Rust
`tensor4all-quanticstransform` crate via C API.

# Usage
```julia
using Tensor4all.QuanticsTransform

op = shift_operator(4, 1)
result = apply(op, mps)
```
"""
module QuanticsTransform

using ..C_API
using ..TreeTN: TreeTensorNetwork

export LinearOperator
export shift_operator, flip_operator, phase_rotation_operator, cumsum_operator, fourier_operator
export apply
export BoundaryCondition, Periodic, Open

# ============================================================================
# Boundary condition enum
# ============================================================================

"""
    BoundaryCondition

Boundary condition for quantics operators.

- `Periodic` (0): Periodic boundary conditions
- `Open` (1): Open boundary conditions
"""
@enum BoundaryCondition begin
    Periodic = 0
    Open = 1
end

# ============================================================================
# LinearOperator type
# ============================================================================

"""
    LinearOperator

A linear operator that can be applied to a tree tensor network (MPS).

Wraps a Rust `LinearOperator` from the quantics transform crate.
Created via operator construction functions like `shift_operator`, `flip_operator`, etc.
"""
mutable struct LinearOperator
    ptr::Ptr{Cvoid}

    function LinearOperator(ptr::Ptr{Cvoid})
        ptr == C_NULL && error("Failed to create LinearOperator: null pointer")
        op = new(ptr)
        finalizer(op) do obj
            C_API.t4a_linop_release(obj.ptr)
        end
        return op
    end
end

# ============================================================================
# Operator construction functions
# ============================================================================

"""
    shift_operator(r::Integer, offset::Integer; bc=Periodic) -> LinearOperator

Create a shift operator: f(x) = g(x + offset) mod 2^r.

# Arguments
- `r`: Number of quantics bits
- `offset`: Shift offset (can be negative)
- `bc`: Boundary condition (`Periodic` or `Open`)
"""
function shift_operator(r::Integer, offset::Integer; bc::BoundaryCondition=Periodic)
    out = Ref{Ptr{Cvoid}}(C_NULL)
    status = C_API.t4a_qtransform_shift(Csize_t(r), Int64(offset), Cint(Int(bc)), out)
    C_API.check_status(status)
    return LinearOperator(out[])
end

"""
    flip_operator(r::Integer; bc=Periodic) -> LinearOperator

Create a flip operator: f(x) = g(2^r - x).

# Arguments
- `r`: Number of quantics bits
- `bc`: Boundary condition (`Periodic` or `Open`)
"""
function flip_operator(r::Integer; bc::BoundaryCondition=Periodic)
    out = Ref{Ptr{Cvoid}}(C_NULL)
    status = C_API.t4a_qtransform_flip(Csize_t(r), Cint(Int(bc)), out)
    C_API.check_status(status)
    return LinearOperator(out[])
end

"""
    phase_rotation_operator(r::Integer, theta::Real) -> LinearOperator

Create a phase rotation operator: f(x) = exp(i*theta*x) * g(x).

# Arguments
- `r`: Number of quantics bits
- `theta`: Phase angle
"""
function phase_rotation_operator(r::Integer, theta::Real)
    out = Ref{Ptr{Cvoid}}(C_NULL)
    status = C_API.t4a_qtransform_phase_rotation(Csize_t(r), Cdouble(theta), out)
    C_API.check_status(status)
    return LinearOperator(out[])
end

"""
    cumsum_operator(r::Integer) -> LinearOperator

Create a cumulative sum operator: y_i = sum_{j<i} x_j.

# Arguments
- `r`: Number of quantics bits
"""
function cumsum_operator(r::Integer)
    out = Ref{Ptr{Cvoid}}(C_NULL)
    status = C_API.t4a_qtransform_cumsum(Csize_t(r), out)
    C_API.check_status(status)
    return LinearOperator(out[])
end

"""
    fourier_operator(r::Integer; forward=true, maxbonddim=0, tolerance=0.0) -> LinearOperator

Create a Quantics Fourier Transform operator.

# Arguments
- `r`: Number of quantics bits
- `forward`: `true` for forward FT, `false` for inverse
- `maxbonddim`: Maximum bond dimension (0 = default of 12)
- `tolerance`: Tolerance for FT construction (0.0 = default of 1e-14)
"""
function fourier_operator(r::Integer; forward::Bool=true, maxbonddim::Integer=0, tolerance::Real=0.0)
    out = Ref{Ptr{Cvoid}}(C_NULL)
    status = C_API.t4a_qtransform_fourier(
        Csize_t(r), Cint(forward ? 1 : 0),
        Csize_t(maxbonddim), Cdouble(tolerance), out
    )
    C_API.check_status(status)
    return LinearOperator(out[])
end

# ============================================================================
# Operator application
# ============================================================================

"""
    apply(op::LinearOperator, state::TreeTensorNetwork; method=:naive, rtol=0.0, maxdim=0) -> TreeTensorNetwork

Apply a linear operator to a tree tensor network (MPS).

# Arguments
- `op`: Linear operator
- `state`: Input MPS/TreeTN
- `method`: Contraction method - `:naive`, `:zipup`, or `:fit`
- `rtol`: Relative tolerance (0.0 = default)
- `maxdim`: Maximum bond dimension (0 = unlimited)
"""
function apply(op::LinearOperator, state::TreeTensorNetwork;
               method::Symbol=:naive, rtol::Real=0.0, maxdim::Integer=0)
    method_int = if method == :naive
        Cint(0)
    elseif method == :zipup
        Cint(1)
    elseif method == :fit
        Cint(2)
    else
        error("Unknown method: $method. Use :naive, :zipup, or :fit")
    end

    out = Ref{Ptr{Cvoid}}(C_NULL)
    status = C_API.t4a_linop_apply(
        op.ptr, state.handle, method_int,
        Cdouble(rtol), Csize_t(maxdim), out
    )
    C_API.check_status(status)

    # Get the number of vertices from the result
    n_out = Ref{Csize_t}(0)
    status = C_API.t4a_treetn_num_vertices(out[], n_out)
    C_API.check_status(status)
    n = Int(n_out[])

    # Result TreeTN has usize node names (0, 1, ..., n-1)
    node_names = collect(0:n-1)
    node_map = Dict{Int, Int}(name => name for name in node_names)
    return TreeTensorNetwork{Int}(out[], node_map, node_names)
end

# ============================================================================
# Display
# ============================================================================

function Base.show(io::IO, op::LinearOperator)
    print(io, "LinearOperator()")
end

# ============================================================================
# Copy / Clone
# ============================================================================

function Base.copy(op::LinearOperator)
    ptr = C_API.t4a_linop_clone(op.ptr)
    return LinearOperator(ptr)
end

function Base.deepcopy(op::LinearOperator)
    ptr = C_API.t4a_linop_clone(op.ptr)
    return LinearOperator(ptr)
end

end # module QuanticsTransform
